File: examples/go/kem_demo.go
// QUAC 100 ML-KEM (Kyber) Key Exchange Demo (Go)
//
// Demonstrates complete key encapsulation mechanism workflow:
// - Key generation
// - Encapsulation (sender side)
// - Decapsulation (receiver side)
// - Shared secret verification
//
// Build: go build -o kem_demo kem_demo.go
// Run:   ./kem_demo [512|768|1024]
//
// Copyright 2025 Dyber, Inc. All Rights Reserved.

package main

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"flag"
	"fmt"
	"os"
)

// Algorithm represents KEM algorithm variants
type Algorithm int

const (
	AlgMLKEM512 Algorithm = iota
	AlgMLKEM768
	AlgMLKEM1024
)

// KEMParams holds algorithm parameters
type KEMParams struct {
	Algorithm Algorithm
	Name      string
	PKSize    int
	SKSize    int
	CTSize    int
	SSSize    int
}

var kemParams = map[int]KEMParams{
	512:  {AlgMLKEM512, "ML-KEM-512", 800, 1632, 768, 32},
	768:  {AlgMLKEM768, "ML-KEM-768", 1184, 2400, 1088, 32},
	1024: {AlgMLKEM1024, "ML-KEM-1024", 1568, 3168, 1568, 32},
}

// SimulatedKEM provides simulated KEM operations
type SimulatedKEM struct {
	params  KEMParams
	lastSS  []byte
}

func NewSimulatedKEM(params KEMParams) *SimulatedKEM {
	return &SimulatedKEM{params: params}
}

func (k *SimulatedKEM) Keygen() (pk, sk []byte, err error) {
	pk = make([]byte, k.params.PKSize)
	sk = make([]byte, k.params.SKSize)
	rand.Read(pk)
	rand.Read(sk)
	return pk, sk, nil
}

func (k *SimulatedKEM) Encaps(pk []byte) (ct, ss []byte, err error) {
	ct = make([]byte, k.params.CTSize)
	ss = make([]byte, k.params.SSSize)
	rand.Read(ct)
	rand.Read(ss)
	k.lastSS = ss // Store for simulation
	return ct, ss, nil
}

func (k *SimulatedKEM) Decaps(ct, sk []byte) (ss []byte, err error) {
	// In simulation, return the same shared secret
	if k.lastSS != nil {
		return k.lastSS, nil
	}
	ss = make([]byte, k.params.SSSize)
	rand.Read(ss)
	return ss, nil
}

func printHex(label string, data []byte, maxBytes int) {
	if maxBytes > len(data) {
		maxBytes = len(data)
	}
	suffix := ""
	if len(data) > maxBytes {
		suffix = "..."
	}
	fmt.Printf("%s (%d bytes):\n  %s%s\n", label, len(data),
		hex.EncodeToString(data[:maxBytes]), suffix)
}

func main() {
	// Parse command line
	level := flag.Int("level", 768, "Security level: 512, 768, or 1024")
	flag.Parse()

	// Get parameters
	params, ok := kemParams[*level]
	if !ok {
		fmt.Fprintf(os.Stderr, "Error: Invalid level %d. Use 512, 768, or 1024.\n", *level)
		os.Exit(1)
	}

	fmt.Println("================================================================")
	fmt.Println("  QUAC 100 ML-KEM Key Exchange Demo (Go)")
	fmt.Printf("  Algorithm: %s (FIPS 203)\n", params.Name)
	fmt.Println("================================================================")
	fmt.Println()

	// Initialize
	kem := NewSimulatedKEM(params)
	fmt.Println("Using software simulator.")
	fmt.Println()

	// =========================================================================
	// Step 1: Key Generation (Receiver - Alice)
	// =========================================================================
	fmt.Println("Step 1: Key Generation (Receiver - Alice)")
	fmt.Println("-" + "--------------------------------------------")
	fmt.Println("Alice generates a keypair to receive encrypted messages.")
	fmt.Println()

	alicePK, aliceSK, err := kem.Keygen()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Key generation failed: %v\n", err)
		os.Exit(1)
	}

	printHex("Alice's Public Key", alicePK, 48)
	fmt.Printf("Alice's Secret Key: %d bytes (kept private)\n\n", len(aliceSK))
	fmt.Println("Alice sends her public key to Bob...")
	fmt.Println()

	// =========================================================================
	// Step 2: Encapsulation (Sender - Bob)
	// =========================================================================
	fmt.Println("Step 2: Encapsulation (Sender - Bob)")
	fmt.Println("-------------------------------------")
	fmt.Println("Bob uses Alice's public key to create:")
	fmt.Println("  - A ciphertext (to send to Alice)")
	fmt.Println("  - A shared secret (kept by Bob)")
	fmt.Println()

	ciphertext, ssBob, err := kem.Encaps(alicePK)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Encapsulation failed: %v\n", err)
		os.Exit(1)
	}

	printHex("Ciphertext", ciphertext, 48)
	printHex("Bob's Shared Secret", ssBob, 32)
	fmt.Println()
	fmt.Println("Bob sends the ciphertext to Alice...")
	fmt.Println()

	// =========================================================================
	// Step 3: Decapsulation (Receiver - Alice)
	// =========================================================================
	fmt.Println("Step 3: Decapsulation (Receiver - Alice)")
	fmt.Println("-----------------------------------------")
	fmt.Println("Alice uses her secret key to recover the shared secret.")
	fmt.Println()

	ssAlice, err := kem.Decaps(ciphertext, aliceSK)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Decapsulation failed: %v\n", err)
		os.Exit(1)
	}

	printHex("Alice's Shared Secret", ssAlice, 32)
	fmt.Println()

	// =========================================================================
	// Step 4: Verification
	// =========================================================================
	fmt.Println("Step 4: Verification")
	fmt.Println("--------------------")

	if bytes.Equal(ssBob, ssAlice) {
		fmt.Println("✓ SUCCESS! Both parties have the same shared secret.")
		fmt.Println("  This secret can now be used as a symmetric encryption key.")
		fmt.Println()
	} else {
		fmt.Println("✗ FAILURE! Shared secrets do not match.")
		os.Exit(1)
	}

	// =========================================================================
	// Summary
	// =========================================================================
	fmt.Println("================================================================")
	fmt.Println("  Key Exchange Complete")
	fmt.Println("================================================================")
	fmt.Printf("Algorithm:      %s\n", params.Name)
	fmt.Printf("Public Key:     %d bytes\n", len(alicePK))
	fmt.Printf("Secret Key:     %d bytes\n", len(aliceSK))
	fmt.Printf("Ciphertext:     %d bytes\n", len(ciphertext))
	fmt.Printf("Shared Secret:  %d bytes (256 bits)\n", len(ssBob))
	fmt.Println()
	fmt.Println("This shared secret provides:")
	fmt.Println("  • Post-quantum security against Shor's algorithm")
	fmt.Println("  • IND-CCA2 security (chosen ciphertext attack resistance)")
	fmt.Println("  • Perfect forward secrecy when used with ephemeral keys")
	fmt.Println("================================================================")

	// Secure cleanup
	for i := range aliceSK {
		aliceSK[i] = 0
	}
	for i := range ssBob {
		ssBob[i] = 0
	}
	for i := range ssAlice {
		ssAlice[i] = 0
	}
}

File: examples/java/KEMDemo.java
/**
 * QUAC 100 ML-KEM (Kyber) Key Exchange Demo (Java)
 * 
 * Demonstrates complete key encapsulation mechanism workflow:
 * - Key generation
 * - Encapsulation (sender side)
 * - Decapsulation (receiver side)
 * - Shared secret verification
 * 
 * Build: javac KEMDemo.java
 * Run:   java KEMDemo [512|768|1024]
 * 
 * Copyright 2025 Dyber, Inc. All Rights Reserved.
 */

import java.security.SecureRandom;
import java.util.Arrays;

/**
 * KEM Algorithm Parameters
 */
class KEMParams {
    public final String name;
    public final int pkSize;
    public final int skSize;
    public final int ctSize;
    public final int ssSize;
    
    public KEMParams(String name, int pkSize, int skSize, int ctSize, int ssSize) {
        this.name = name;
        this.pkSize = pkSize;
        this.skSize = skSize;
        this.ctSize = ctSize;
        this.ssSize = ssSize;
    }
    
    public static KEMParams get(int level) {
        switch (level) {
            case 512:  return new KEMParams("ML-KEM-512", 800, 1632, 768, 32);
            case 768:  return new KEMParams("ML-KEM-768", 1184, 2400, 1088, 32);
            case 1024: return new KEMParams("ML-KEM-1024", 1568, 3168, 1568, 32);
            default:   return new KEMParams("ML-KEM-768", 1184, 2400, 1088, 32);
        }
    }
}

/**
 * KEM Key Pair
 */
class KEMKeyPair {
    public final byte[] publicKey;
    public final byte[] secretKey;
    
    public KEMKeyPair(byte[] publicKey, byte[] secretKey) {
        this.publicKey = publicKey;
        this.secretKey = secretKey;
    }
    
    public void clear() {
        Arrays.fill(secretKey, (byte) 0);
    }
}

/**
 * Encapsulation Result
 */
class EncapsResult {
    public final byte[] ciphertext;
    public final byte[] sharedSecret;
    
    public EncapsResult(byte[] ciphertext, byte[] sharedSecret) {
        this.ciphertext = ciphertext;
        this.sharedSecret = sharedSecret;
    }
    
    public void clear() {
        Arrays.fill(sharedSecret, (byte) 0);
    }
}

/**
 * Simulated KEM Operations
 */
class SimulatedKEM {
    private final KEMParams params;
    private final SecureRandom random = new SecureRandom();
    private byte[] lastSS = null;
    
    public SimulatedKEM(KEMParams params) {
        this.params = params;
    }
    
    public KEMKeyPair keygen() {
        byte[] pk = new byte[params.pkSize];
        byte[] sk = new byte[params.skSize];
        random.nextBytes(pk);
        random.nextBytes(sk);
        return new KEMKeyPair(pk, sk);
    }
    
    public EncapsResult encaps(byte[] pk) {
        byte[] ct = new byte[params.ctSize];
        byte[] ss = new byte[params.ssSize];
        random.nextBytes(ct);
        random.nextBytes(ss);
        lastSS = ss.clone(); // Store for simulation
        return new EncapsResult(ct, ss);
    }
    
    public byte[] decaps(byte[] ct, byte[] sk) {
        // Return stored secret for simulation
        if (lastSS != null) {
            return lastSS.clone();
        }
        byte[] ss = new byte[params.ssSize];
        random.nextBytes(ss);
        return ss;
    }
}

/**
 * Main KEM Demo
 */
public class KEMDemo {
    
    private static String toHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
    
    private static void printHex(String label, byte[] data, int maxBytes) {
        String hex = toHex(Arrays.copyOf(data, Math.min(data.length, maxBytes)));
        String suffix = data.length > maxBytes ? "..." : "";
        System.out.println(label + " (" + data.length + " bytes):");
        System.out.println("  " + hex + suffix);
    }
    
    public static void main(String[] args) {
        // Parse command line
        int level = 768;
        if (args.length > 0) {
            try {
                level = Integer.parseInt(args[0]);
                if (level != 512 && level != 768 && level != 1024) {
                    System.err.println("Error: Invalid level. Use 512, 768, or 1024.");
                    System.exit(1);
                }
            } catch (NumberFormatException e) {
                System.err.println("Usage: java KEMDemo [512|768|1024]");
                System.exit(1);
            }
        }
        
        KEMParams params = KEMParams.get(level);
        
        System.out.println("================================================================");
        System.out.println("  QUAC 100 ML-KEM Key Exchange Demo (Java)");
        System.out.println("  Algorithm: " + params.name + " (FIPS 203)");
        System.out.println("================================================================");
        System.out.println();
        
        // Initialize
        SimulatedKEM kem = new SimulatedKEM(params);
        System.out.println("Using software simulator.");
        System.out.println();
        
        // =====================================================================
        // Step 1: Key Generation (Receiver - Alice)
        // =====================================================================
        System.out.println("Step 1: Key Generation (Receiver - Alice)");
        System.out.println("------------------------------------------");
        System.out.println("Alice generates a keypair to receive encrypted messages.");
        System.out.println();
        
        KEMKeyPair aliceKP = kem.keygen();
        
        printHex("Alice's Public Key", aliceKP.publicKey, 48);
        System.out.println("Alice's Secret Key: " + aliceKP.secretKey.length + " bytes (kept private)");
        System.out.println();
        System.out.println("Alice sends her public key to Bob...");
        System.out.println();
        
        // =====================================================================
        // Step 2: Encapsulation (Sender - Bob)
        // =====================================================================
        System.out.println("Step 2: Encapsulation (Sender - Bob)");
        System.out.println("-------------------------------------");
        System.out.println("Bob uses Alice's public key to create:");
        System.out.println("  - A ciphertext (to send to Alice)");
        System.out.println("  - A shared secret (kept by Bob)");
        System.out.println();
        
        EncapsResult bobResult = kem.encaps(aliceKP.publicKey);
        
        printHex("Ciphertext", bobResult.ciphertext, 48);
        printHex("Bob's Shared Secret", bobResult.sharedSecret, 32);
        System.out.println();
        System.out.println("Bob sends the ciphertext to Alice...");
        System.out.println();
        
        // =====================================================================
        // Step 3: Decapsulation (Receiver - Alice)
        // =====================================================================
        System.out.println("Step 3: Decapsulation (Receiver - Alice)");
        System.out.println("-----------------------------------------");
        System.out.println("Alice uses her secret key to recover the shared secret.");
        System.out.println();
        
        byte[] aliceSS = kem.decaps(bobResult.ciphertext, aliceKP.secretKey);
        
        printHex("Alice's Shared Secret", aliceSS, 32);
        System.out.println();
        
        // =====================================================================
        // Step 4: Verification
        // =====================================================================
        System.out.println("Step 4: Verification");
        System.out.println("--------------------");
        
        if (Arrays.equals(bobResult.sharedSecret, aliceSS)) {
            System.out.println("✓ SUCCESS! Both parties have the same shared secret.");
            System.out.println("  This secret can now be used as a symmetric encryption key.");
            System.out.println();
        } else {
            System.out.println("✗ FAILURE! Shared secrets do not match.");
            System.exit(1);
        }
        
        // =====================================================================
        // Summary
        // =====================================================================
        System.out.println("================================================================");
        System.out.println("  Key Exchange Complete");
        System.out.println("================================================================");
        System.out.println("Algorithm:      " + params.name);
        System.out.println("Public Key:     " + aliceKP.publicKey.length + " bytes");
        System.out.println("Secret Key:     " + aliceKP.secretKey.length + " bytes");
        System.out.println("Ciphertext:     " + bobResult.ciphertext.length + " bytes");
        System.out.println("Shared Secret:  " + bobResult.sharedSecret.length + " bytes (256 bits)");
        System.out.println();
        System.out.println("This shared secret provides:");
        System.out.println("  • Post-quantum security against Shor's algorithm");
        System.out.println("  • IND-CCA2 security (chosen ciphertext attack resistance)");
        System.out.println("  • Perfect forward secrecy when used with ephemeral keys");
        System.out.println("================================================================");
        
        // Secure cleanup
        aliceKP.clear();
        bobResult.clear();
        Arrays.fill(aliceSS, (byte) 0);
    }
}